### PR TITLE
feat: Support configuring fixed values automatically from Model

### PR DESCRIPTION
**Pull request by kratsg**
# Description

Replaces #846.

Provides support for automatically passing through the fixed parameters from the model configuration into all of the minimization/inference that we do in the `pyhf` code-base, namely:

- `pyhf.infer.hypotest`
- `pyhf.infer.calculators.generate_asimov_data`
- `pyhf.infer.calculators.AsymptoticCalculators`
- `pyhf.infer.mle.fit`
- `pyhf.infer.mle.fixed_poi_fit`
- `pyhf.infer.test_statistics._qmu_like`
- `pyhf.infer.test_statistics._tmu_like`
- `pyhf.infer.test_statistics.qmu`
- `pyhf.infer.test_statistics.qmu_tilde`
- `pyhf.infer.test_statistics.tmu`
- `pyhf.infer.test_statistics.tmu_tilde`

We will pass in `fixed_params` all the way through to the optimization which will stitch together the two if need-be.

This is a two-fold feature that needs to be refactored/be more elegant a bit later. `fixed_vals` are things that are a list of tuples like so: `[(parameter_index, parameter_value), ...]`. `fixed_params` are what comes from the `pdf.config.suggested_fixed()`.

- in cases where `fixed_params` is optional, the model fixed parameters are retrieved
- in cases where `fixed_params` is not optional, nothing is done (a pass-through)

This tries to model the behavior we have identically for `init_pars` and `par_bounds` everywhere.

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR

```
* Add support for automatically passing through fixed parameters from the model configuration into all of the minimization/inference API
* Add tests for fixed values

Co-authored-by: Carlo Gottardo <carlo.gottardo91@gmail.com>
```

### CODE REVIEW

File: src/pyhf/infer/__init__.py [hunk 1]
```diff
@@ -118,15 +126,31 @@ def hypotest(
...
+        for index, (init, is_fixed) in enumerate(
+            zip(init_pars, pdf.config.suggested_fixed())
+        )
+        if is_fixed
+    ]
+    # add user-defined ones at the end
+    fixed_vals = model_fixed_vals + (fixed_vals or [])
+
+    # de-dupe and use last-appended result for each index
+    fixed_vals = list(dict(fixed_vals).items())
```
Comment (matthewfeickert) [inline src/pyhf/infer/__init__.py:141]: @kratsg can you add the simple example you showed me of this to the PR body for posterity?

```python
>>> foo = [('a', 1), ('a', 2)]
>>> list(dict(foo))
[('a', 2)]
```

File: src/pyhf/infer/__init__.py [hunk 2]
```diff
@@ -118,15 +126,31 @@ def hypotest(
...
+        data, pdf, init_pars, par_bounds, fixed_vals, qtilde=qtilde
+    )
     teststat = calc.teststatistic(poi_test)
     sig_plus_bkg_distribution, b_only_distribution = calc.distributions(poi_test)
 
     CLsb = sig_plus_bkg_distribution.pvalue(teststat)
     CLb = b_only_distribution.pvalue(teststat)
     CLs = CLsb / CLb
+
+    tensorlib, _ = get_backend()
```
Comment (matthewfeickert) [inline src/pyhf/infer/__init__.py:153]: Why is this setting of `tensorlib` needed? Sorry, just not immediately obvious.

File: src/pyhf/infer/calculators.py [hunk 3]
```diff
@@ -22,12 +22,15 @@ def generate_asimov_data(asimov_mu, data, pdf, init_pars, par_bounds):
         pdf (~pyhf.pdf.Model): The statistical model adhering to the schema ``model.json``.
         init_pars (`tensor`): The initial parameter values to be used for fitting.
         par_bounds (`tensor`): The parameter value bounds to be used for fitting.
+        fixed_vals (`tensor`): Parameters to be held constant in the fit.
```
Comment (matthewfeickert) [inline src/pyhf/infer/calculators.py:25]: We should probably harmonize this docstring across the whole PR. I don't have strong feelings one way or another on it.

File: src/pyhf/infer/calculators.py [hunk 4]
```diff
@@ -127,6 +139,8 @@ def __init__(self, data, pdf, init_pars=None, par_bounds=None, qtilde=False):
             pdf (~pyhf.pdf.Model): The statistical model adhering to the schema ``model.json``.
             init_pars (`tensor`): The initial parameter values to be used for fitting.
             par_bounds (`tensor`): The parameter value bounds to be used for fitting.
+            fixed_params (`tensor`): A list of booleans for each parameter on whether it should be fixed in the fit or not.
```
Comment (matthewfeickert) [inline src/pyhf/infer/calculators.py:142]: Is this a tensor or a list?

File: src/pyhf/infer/calculators.py [hunk 5]
```diff
@@ -137,6 +151,22 @@ def __init__(self, data, pdf, init_pars=None, par_bounds=None, qtilde=False):
...
+            )
+            if is_fixed
+        ]
+        # add user-defined ones at the end
+        fixed_vals = model_fixed_vals + (fixed_vals or [])
+
+        # de-dupe and use last-appended result for each index
+        fixed_vals = list(dict(fixed_vals).items())
+
+        self.fixed_vals = fixed_vals
```
Comment (matthewfeickert) [inline src/pyhf/infer/calculators.py:169]:
```suggestion
        # de-dupe and use last-appended result for each index
        self.fixed_vals = list(dict(fixed_vals).items())
```

Maybe just save a few lines.

File: src/pyhf/infer/test_statistics.py [hunk 6]
```diff
@@ -82,7 +84,8 @@ def qmu(mu, data, pdf, init_pars, par_bounds):
         >>> init_pars = model.config.suggested_init()
         >>> par_bounds = model.config.suggested_bounds()
         >>> par_bounds[model.config.poi_index] = [-10.0, 10.0]
-        >>> pyhf.infer.test_statistics.qmu(test_mu, data, model, init_pars, par_bounds)
+        >>> fixed_vals = []
+        >>> pyhf.infer.test_statistics.qmu(test_mu, data, model, init_pars, par_bounds, [])
```
Comment (matthewfeickert) [inline src/pyhf/infer/test_statistics.py:88]:
```suggestion
        >>> pyhf.infer.test_statistics.qmu(test_mu, data, model, init_pars, par_bounds, fixed_vals)
```

As `fixed_vals=[ ]` is used above, we should use the `fixed_vals` variable here.

File: src/pyhf/infer/test_statistics.py [hunk 7]
```diff
@@ -146,7 +150,7 @@ def qmu_tilde(mu, data, pdf, init_pars, par_bounds):
         >>> test_mu = 1.0
         >>> init_pars = model.config.suggested_init()
         >>> par_bounds = model.config.suggested_bounds()
-        >>> pyhf.infer.test_statistics.qmu_tilde(test_mu, data, model, init_pars, par_bounds)
+        >>> pyhf.infer.test_statistics.qmu_tilde(test_mu, data, model, init_pars, par_bounds, [])
```
Comment (matthewfeickert) [inline src/pyhf/infer/test_statistics.py:153]:
```suggestion
        >>> fixed_vals = []
        >>> pyhf.infer.test_statistics.qmu_tilde(test_mu, data, model, init_pars, par_bounds, fixed_vals)
```

File: src/pyhf/infer/test_statistics.py [hunk 8]
```diff
@@ -199,7 +204,8 @@ def tmu(mu, data, pdf, init_pars, par_bounds):
         >>> init_pars = model.config.suggested_init()
         >>> par_bounds = model.config.suggested_bounds()
         >>> par_bounds[model.config.poi_index] = [-10.0, 10.0]
-        >>> pyhf.infer.test_statistics.tmu(test_mu, data, model, init_pars, par_bounds)
+        >>> fixed_vals = []
+        >>> pyhf.infer.test_statistics.tmu(test_mu, data, model, init_pars, par_bounds, [])
```
Comment (matthewfeickert) [inline src/pyhf/infer/test_statistics.py:208]:
```suggestion
        >>> pyhf.infer.test_statistics.tmu(test_mu, data, model, init_pars, par_bounds, fixed_vals)
```

File: src/pyhf/infer/test_statistics.py [hunk 9]
```diff
@@ -257,7 +264,8 @@ def tmu_tilde(mu, data, pdf, init_pars, par_bounds):
         >>> test_mu = 1.0
         >>> init_pars = model.config.suggested_init()
         >>> par_bounds = model.config.suggested_bounds()
-        >>> pyhf.infer.test_statistics.tmu_tilde(test_mu, data, model, init_pars, par_bounds)
+        >>> fixed_vals = []
+        >>> pyhf.infer.test_statistics.tmu_tilde(test_mu, data, model, init_pars, par_bounds, [])
```
Comment (matthewfeickert) [inline src/pyhf/infer/test_statistics.py:268]:
```suggestion
        >>> pyhf.infer.test_statistics.tmu_tilde(test_mu, data, model, init_pars, par_bounds, fixed_vals)
```

File: tests/test_teststats.py [hunk 10]
```diff
@@ -12,7 +12,7 @@ def test_qmu(caplog):
     par_bounds = model.config.suggested_bounds()
 
     with caplog.at_level(logging.WARNING, "pyhf.infer.test_statistics"):
-        pyhf.infer.test_statistics.qmu(mu, data, model, init_pars, par_bounds)
+        pyhf.infer.test_statistics.qmu(mu, data, model, init_pars, par_bounds, [])
```
Comment (matthewfeickert) [inline tests/test_teststats.py:15]:
```suggestion
        pyhf.infer.test_statistics.qmu(mu, data, model, init_pars, par_bounds, fixed_vals)
```

To try and keep this in a similar style as the rest of the function call, and also readable, what about adding a

```python
fixed_vals = []
```
just below `par_bounds = model.config.suggested_bounds()` above and then using it here and then doing the same for the tests below?

**Review by matthewfeickert — COMMENTED** [review]
@kratsg Core wise this is great. Thanks for stepping up and knocking this out this week. Just some readability suggestions and questions for the most part.

**Comment by alexander-held** [comment]
Suggestion: For the minuit optimizer, the initial step size is set in `minuit_optimizer._get_minimizer`. For a fixed Gaussian-constrained nuisance parameter with default bounds `[-5, 5]`, this is `0.01` with the default 1000 steps. This will be the parameter uncertainty reported by minuit after the fit. Given that it does not makes sense to have an uncertainty for a fixed parameter, it might be nice to set the entries of `step_sizes` corresponding to fixed parameters to `0` instead.

File: src/pyhf/infer/calculators.py (hunk 4 above)
Comment (kratsg) [reply src/pyhf/infer/calculators.py]: can be either.

File: src/pyhf/infer/__init__.py (hunk 2 above)
Comment (kratsg) [reply src/pyhf/infer/__init__.py:141]: I moved it further down to where it's needed.

**Review by matthewfeickert — APPROVED** [review]
@kanishk16 LGTM for the original scope. As this is a release PR if we can get @lukasheinrich's review as well that would be good.

I haven't looked at @alexander-held's suggestion in terms of scope yet, so should we try to get that in here or is that beyond scope and should be its own Issue + PR? Also, would be good to understand how the changes in 0ad8a7d43273208269f201f8a333f3f22007f390 managed to break everything if there's time. :?

**Comment by kratsg** [comment]
> Suggestion: For the minuit optimizer, the initial step size is set in `minuit_optimizer._get_minimizer`. For a fixed Gaussian-constrained nuisance parameter with default bounds `[-5, 5]`, this is `0.01` with the default 1000 steps. This will be the parameter uncertainty reported by minuit after the fit. Given that it does not make sense to have an uncertainty for a fixed parameter, it might be nice to set the entries of `step_sizes` corresponding to fixed parameters to `0` instead.

This is done in #1054.

**Review by lukasheinrich — APPROVED** [review]

**Pull request merged**